log/context.go

```diff
@@ -15,18 +15,16 @@ type contextKeyType struct{}
 var contextKey contextKeyType
 
 func init() {
+	copyContext := func(dst, src context.Context) context.Context {
+		if logger, ok := src.Value(contextKey).(*zap.SugaredLogger); ok && logger != nil {
+			dst = context.WithValue(dst, contextKey, logger)
+		}
+		return dst
+	}
 	detach.Register(detach.Hooks{
-		Inline: func(dst, src context.Context) context.Context {
-			if logger, ok := src.Value(contextKey).(*zap.SugaredLogger); ok && logger != nil {
-				dst = context.WithValue(dst, contextKey, logger)
-			}
-			return dst
-		},
+		Inline: copyContext,
 		Async: func(dst, src context.Context, next func(ctx context.Context)) {
-			if logger, ok := src.Value(contextKey).(*zap.SugaredLogger); ok && logger != nil {
-				dst = context.WithValue(dst, contextKey, logger)
-			}
-			next(dst)
+			next(copyContext(dst, src))
```

Comment:
should we make this the default behavior if async is nil?

Reply:
default behavior could cause footgun (as packages need to put in an explicit do-nothing `Async` here if they can't copy anything in `Async`), but I proposed to provide an implementation packages can directly reuse in https://github.com/reddit/baseplate.go/pull/446#discussion_r762344054

Reply:
but default behavior does have the benefit of not adding hook to the function signature, so it might be worth it :)

Reply:
> should we make this the default behavior if async is nil?

I don't think so, like yuxuan said, that might not be safe. I really want this part to be explicit since it's really only used by a small number of libraries in framework code so I would rather force that part to be as direct as possible which is what I like about the current interface it provides.